diff --git a/vllm/config.py b/vllm/config.py
index 5b255e210..6c85c235c 100644
--- a/vllm/config.py
+++ b/vllm/config.py
@@ -2301,6 +2301,12 @@ class SpeculativeConfig:
     """If set to True, token log probabilities are not returned during
     speculative decoding. If set to False, token log probabilities are returned
     according to the log probability settings in SamplingParams."""
+    enable_adaptive: bool = False
+    """Enable adaptive speculative decoding to adjust the number of
+    speculative tokens based on draft model accuracy."""
+    min_num_speculative_tokens: int = 1
+    """Minimum number of speculative tokens when adaptive speculative decoding
+    is enabled."""
 
     # Draft model configuration
     quantization: Optional[QuantizationMethods] = None
diff --git a/vllm/v1/core/sched/scheduler.py b/vllm/v1/core/sched/scheduler.py
index b0a6841f3..aba9fb9a3 100644
--- a/vllm/v1/core/sched/scheduler.py
+++ b/vllm/v1/core/sched/scheduler.py
@@ -864,7 +864,7 @@ class Scheduler(SchedulerInterface):
                 request.num_computed_tokens -= num_tokens_rejected
                 spec_decoding_stats = self.make_spec_decoding_stats(
                     spec_decoding_stats,
-                    num_draft_tokens=len(scheduled_spec_token_ids) * num_steps,
+                    num_draft_tokens=self.num_spec_tokens * num_steps,
                     num_accepted_tokens=len(generated_token_ids) - num_steps)
 
             cached_encoder_input_ids = (
@@ -1129,7 +1129,9 @@ class Scheduler(SchedulerInterface):
                 request.append_output_token_ids(request.sampling_params.extra_args['kv_transfer_params']['prefilled_token'])
         
         # due to no speculative trans from prefill node, push one manually
-        if self.vllm_config.speculative_config is not None and (self.vllm_config.kv_transfer_config is None or self.vllm_config.kv_transfer_config.kv_role != "kv_both"):
+        if (self.vllm_config.speculative_config is not None and
+            (self.vllm_config.kv_transfer_config is None or self.vllm_config.kv_transfer_config.kv_role != "kv_both") and
+            not self.vllm_config.speculative_config.enable_adaptive):
             request.spec_token_ids.extend([0] * self.vllm_config.speculative_config.num_speculative_tokens)
         # Update the request state for scheduling.
         request.num_computed_tokens = request.num_tokens - 1
