diff --git a/vllm/entrypoints/openai/serving_chat.py b/vllm/entrypoints/openai/serving_chat.py
index 976a4163f..20c9f8056 100755
--- a/vllm/entrypoints/openai/serving_chat.py
+++ b/vllm/entrypoints/openai/serving_chat.py
@@ -468,6 +468,7 @@ class OpenAIServingChat(OpenAIServing):
         finish_reason_sent = [False] * num_choices
         num_prompt_tokens = 0
         num_cached_tokens = None
+        cached_rate = None
 
         # turn id for kimi
         turn_id = 0
@@ -918,10 +919,10 @@ class OpenAIServingChat(OpenAIServing):
                     # handle usage stats if requested & if continuous
                     if include_continuous_usage:
                         completion_tokens = previous_num_tokens[i]
-                        if request.kv_transfer_params and request.kv_transfer_params["num_cached_tokens"]:
+                        if request.kv_transfer_params and "num_cached_tokens" in request.kv_transfer_params and \
+                                request.kv_transfer_params["num_cached_tokens"] and num_prompt_tokens:
                             num_cached_tokens = request.kv_transfer_params["num_cached_tokens"]
-                        cached_rate = None if num_prompt_tokens == 0 else round(
-                            num_cached_tokens / num_prompt_tokens, 4)
+                            cached_rate = round(num_cached_tokens / num_prompt_tokens, 4)
                         chunk.usage = UsageInfo(
                             prompt_tokens=num_prompt_tokens,
                             completion_tokens=completion_tokens,
@@ -947,10 +948,10 @@ class OpenAIServingChat(OpenAIServing):
                         final_usage.prompt_tokens_details = PromptTokenUsageInfo(
                             cached_tokens=0)
                     else:
-                        if request.kv_transfer_params and request.kv_transfer_params["num_cached_tokens"]:
+                        if request.kv_transfer_params and "num_cached_tokens" in request.kv_transfer_params and \
+                                request.kv_transfer_params["num_cached_tokens"] and num_prompt_tokens:
                             num_cached_tokens = request.kv_transfer_params["num_cached_tokens"]
-                        cached_rate = None if num_prompt_tokens == 0 else round(
-                            num_cached_tokens / num_prompt_tokens, 4)
+                            cached_rate = round(num_cached_tokens / num_prompt_tokens, 4)
                         final_usage.prompt_tokens_details = PromptTokenUsageInfo(
                             cached_tokens=num_cached_tokens,
                             cached_rate = cached_rate)
@@ -1223,7 +1224,7 @@ class OpenAIServingChat(OpenAIServing):
                 usage.prompt_tokens_details = PromptTokenUsageInfo(
                     cached_tokens=0)
             else:
-                cached_rate = None if num_prompt_tokens == 0 else round(
+                cached_rate = None if not num_prompt_tokens else round(
                     final_res.num_cached_tokens / num_prompt_tokens, 4)
                 usage.prompt_tokens_details = PromptTokenUsageInfo(
                     cached_tokens=final_res.num_cached_tokens,
diff --git a/vllm/entrypoints/openai/serving_completion.py b/vllm/entrypoints/openai/serving_completion.py
index a1930c244..e608c6bc2 100755
--- a/vllm/entrypoints/openai/serving_completion.py
+++ b/vllm/entrypoints/openai/serving_completion.py
@@ -329,6 +329,7 @@ class OpenAIServingCompletion(OpenAIServing):
         has_echoed = [False] * num_choices * num_prompts
         num_prompt_tokens = [0] * num_prompts
         num_cached_tokens = 0
+        cached_rate = None
 
         stream_options = request.stream_options
         if stream_options:
@@ -424,10 +425,10 @@ class OpenAIServingCompletion(OpenAIServing):
                     if include_continuous_usage:
                         prompt_tokens = num_prompt_tokens[prompt_idx]
                         completion_tokens = previous_num_tokens[i]
-                        if request.kv_transfer_params and request.kv_transfer_params["num_cached_tokens"]:
+                        if request.kv_transfer_params and "num_cached_tokens" in request.kv_transfer_params and \
+                                request.kv_transfer_params["num_cached_tokens"] is not None and prompt_tokens:
                             num_cached_tokens = request.kv_transfer_params["num_cached_tokens"]
-                        cached_rate = None if prompt_tokens == 0 else round(
-                            num_cached_tokens / prompt_tokens, 4)
+                            cached_rate = round(num_cached_tokens / prompt_tokens, 4)
                         chunk.usage = UsageInfo(
                             prompt_tokens=prompt_tokens,
                             completion_tokens=completion_tokens,
@@ -442,10 +443,10 @@ class OpenAIServingCompletion(OpenAIServing):
 
             total_prompt_tokens = sum(num_prompt_tokens)
             total_completion_tokens = sum(previous_num_tokens)
-            if request.kv_transfer_params and request.kv_transfer_params["num_cached_tokens"]:
+            if request.kv_transfer_params and "num_cached_tokens" in request.kv_transfer_params and \
+                    request.kv_transfer_params["num_cached_tokens"] is not None and total_prompt_tokens:
                 num_cached_tokens = request.kv_transfer_params["num_cached_tokens"]
-            cached_rate = None if total_prompt_tokens == 0 else round(
-                num_cached_tokens / total_prompt_tokens, 4)
+                cached_rate = round(num_cached_tokens / total_prompt_tokens, 4)
             final_usage_info = UsageInfo(
                 prompt_tokens=total_prompt_tokens,
                 completion_tokens=total_completion_tokens,
@@ -586,7 +587,7 @@ class OpenAIServingCompletion(OpenAIServing):
                     final_res.kv_transfer_params["stop_reasons"] = [
                         output.stop_reason for output in final_res.outputs
                     ]
-        cached_rate = None if prompt_token_len == 0 else round(
+        cached_rate = None if not prompt_token_len else round(
             num_cached_tokens / prompt_token_len, 4)
         usage = UsageInfo(
             prompt_tokens=num_prompt_tokens,
