diff --git a/vllm/entrypoints/openai/protocol.py b/vllm/entrypoints/openai/protocol.py
index cbadb080f..d60ac404e 100644
--- a/vllm/entrypoints/openai/protocol.py
+++ b/vllm/entrypoints/openai/protocol.py
@@ -1270,6 +1270,8 @@ class CompletionResponseChoice(OpenAIBaseModel):
             "including encountering the EOS token"),
     )
     prompt_logprobs: Optional[list[Optional[dict[int, Logprob]]]] = None
+    routed_experts: Optional[dict[str, Any]] = Field(
+        default=None, description="Expert ids parameters.")
 
 
 class CompletionResponse(OpenAIBaseModel):
@@ -1295,6 +1297,8 @@ class CompletionResponseStreamChoice(OpenAIBaseModel):
             "to stop, None if the completion finished for some other reason "
             "including encountering the EOS token"),
     )
+    routed_experts: Optional[dict[str, Any]] = Field(
+        default=None, description="Expert ids parameters.")
 
 
 class CompletionStreamResponse(OpenAIBaseModel):
@@ -1460,6 +1464,8 @@ class ChatCompletionResponseChoice(OpenAIBaseModel):
     finish_reason: Optional[str] = "stop"
     # not part of the OpenAI spec but included in vLLM for legacy reasons
     stop_reason: Optional[Union[int, str]] = None
+    routed_experts: Optional[dict[str, Any]] = Field(
+        default=None, description="Expert ids parameters.")
 
 
 class ChatCompletionResponse(OpenAIBaseModel):
@@ -1487,6 +1493,8 @@ class ChatCompletionResponseStreamChoice(OpenAIBaseModel):
     logprobs: Optional[ChatCompletionLogProbs] = None
     finish_reason: Optional[str] = None
     stop_reason: Optional[Union[int, str]] = None
+    routed_experts: Optional[dict[str, Any]] = Field(
+        default=None, description="Expert ids parameters.")
 
 
 class ChatCompletionStreamResponse(OpenAIBaseModel):
diff --git a/vllm/entrypoints/openai/serving_chat.py b/vllm/entrypoints/openai/serving_chat.py
index 37d5e1746..a5569c508 100755
--- a/vllm/entrypoints/openai/serving_chat.py
+++ b/vllm/entrypoints/openai/serving_chat.py
@@ -300,6 +300,7 @@ class OpenAIServingChat(OpenAIServing):
                             request.kv_transfer_params.pop("prefilled_token", None)
                         else:
                             engine_prompt["prefilled_token_ids"] = request.kv_transfer_params["prefilled_token"]
+                            engine_prompt["logprobs"] = request.kv_transfer_params["logprobs"]
                             engine_prompt["prefilled_texts"] = delta_text
                             engine_prompt["max_tokens"] = min(request.max_tokens or self.max_model_len, self.max_model_len-len(engine_prompt["prompt_token_ids"]))
                             request.kv_transfer_params["max_tokens"] = engine_prompt["max_tokens"]
@@ -1060,6 +1061,23 @@ class OpenAIServingChat(OpenAIServing):
                         created=created_time,
                         choices=[choice_data],
                         model=model_name)
+                    
+                    expert_ids = getattr(res, 'routed_experts', None)
+                    if expert_ids is not None and expert_ids.shape[0] != 0:
+                        chunk.choices[0].routed_experts = {}
+                        self.add_ndarray_info_to_dict(expert_ids, chunk.choices[0].routed_experts)  
+                    elif request.kv_transfer_params:
+                        required_keys = [
+                                "routed_experts_shape",
+                                "routed_experts_dtype",
+                                "routed_experts_str_len",
+                                "routed_experts_str"
+                            ]
+                        if all(key in request.kv_transfer_params for key in required_keys):
+                            chunk.choices[0].routed_experts = {
+                                key: request.kv_transfer_params[key]
+                                for key in required_keys
+                            }
 
                     # handle usage stats if requested & if continuous
                     if include_continuous_usage:
@@ -1165,6 +1183,7 @@ class OpenAIServingChat(OpenAIServing):
         if reuse_prefilled_tokens:
             if final_res.kv_transfer_params:
                 final_res.kv_transfer_params["prefilled_token"] = [final_res.outputs[0].token_ids[0]]
+                final_res.kv_transfer_params["logprobs"] = final_res.outputs[0].logprobs
 
         choices: list[ChatCompletionResponseChoice] = []
 
@@ -1363,6 +1382,19 @@ class OpenAIServingChat(OpenAIServing):
                 finish_reason="tool_calls" if auto_tools_called else
                 output.finish_reason if output.finish_reason else "stop",
                 stop_reason=output.stop_reason)
+
+            if os.getenv('ROLE', None) == "prefill":
+                experts_id_array = getattr(final_res, 'routed_experts', None)
+                if experts_id_array is not None:
+                    self.add_ndarray_info_to_dict(experts_id_array, final_res.kv_transfer_params)
+            else:
+                kv_transfer_params = getattr(request, 'kv_transfer_params', {}) or {}
+                res_expert_ids = getattr(final_res, 'routed_experts', None)
+                if res_expert_ids is not None:
+                    res_expert_ids = self.concatenate_dict_and_ndarray(kv_transfer_params, res_expert_ids)
+                    choice_data.routed_experts = {}
+                    self.add_ndarray_info_to_dict(res_expert_ids, choice_data.routed_experts)
+
             choices.append(choice_data)
 
         if request.echo:
diff --git a/vllm/entrypoints/openai/serving_completion.py b/vllm/entrypoints/openai/serving_completion.py
index e43a96d51..8cb73ee59 100755
--- a/vllm/entrypoints/openai/serving_completion.py
+++ b/vllm/entrypoints/openai/serving_completion.py
@@ -417,6 +417,24 @@ class OpenAIServingCompletion(OpenAIServing):
                                 stop_reason=stop_reason,
                             )
                         ])
+
+                    expert_ids = getattr(res, 'routed_experts', None)
+                    if expert_ids is not None and expert_ids.shape[0] != 0:
+                        chunk.choices[0].routed_experts = {}
+                        self.add_ndarray_info_to_dict(expert_ids, chunk.choices[0].routed_experts)
+                    elif request.kv_transfer_params:
+                        required_keys = [
+                            "routed_experts_shape",
+                            "routed_experts_dtype",
+                            "routed_experts_str_len",
+                            "routed_experts_str"
+                        ]
+                        if all(key in request.kv_transfer_params for key in required_keys):
+                            chunk.choices[0].routed_experts = {
+                                key: request.kv_transfer_params[key]
+                                for key in required_keys
+                            }
+
                     if include_continuous_usage:
                         prompt_tokens = num_prompt_tokens[prompt_idx]
                         completion_tokens = previous_num_tokens[i]
@@ -549,6 +567,19 @@ class OpenAIServingCompletion(OpenAIServing):
                     stop_reason=output.stop_reason,
                     prompt_logprobs=final_res.prompt_logprobs,
                 )
+
+                if os.getenv('ROLE', None) == "prefill":
+                    experts_id_array = getattr(final_res, 'routed_experts', None)
+                    if experts_id_array is not None:
+                        self.add_ndarray_info_to_dict(experts_id_array, final_res.kv_transfer_params)
+                else:
+                    kv_transfer_params = getattr(request, 'kv_transfer_params', {}) or {}
+                    res_expert_ids = getattr(final_res, 'routed_experts', None)
+                    if res_expert_ids is not None:
+                        res_expert_ids = self.concatenate_dict_and_ndarray(kv_transfer_params, res_expert_ids)
+                        choice_data.routed_experts = {}
+                        self.add_ndarray_info_to_dict(res_expert_ids, choice_data.routed_experts)
+
                 choices.append(choice_data)
 
                 num_generated_tokens += len(output.token_ids)
diff --git a/vllm/v1/engine/async_llm.py b/vllm/v1/engine/async_llm.py
index 160557186..43dd27980 100644
--- a/vllm/v1/engine/async_llm.py
+++ b/vllm/v1/engine/async_llm.py
@@ -305,7 +305,7 @@ class AsyncLLM(EngineClient):
                                 prompt=None, finished=False, prompt_logprobs=None,
                                 prompt_token_ids=prompt["prompt_token_ids"],
                                 outputs=[CompletionOutput(index=0,
-                                    cumulative_logprob=None, logprobs=None,
+                                    cumulative_logprob=None, logprobs=prompt.get('logorobs', None),
                                     text= prompt["prefilled_texts"],
                                     token_ids=prompt["prefilled_token_ids"])])
                     else:
diff --git a/vllm/entrypoints/openai/serving_engine.py b/vllm/entrypoints/openai/serving_engine.py
index 5c58c2e64..d05ed1549 100644
--- a/vllm/entrypoints/openai/serving_engine.py
+++ b/vllm/entrypoints/openai/serving_engine.py
@@ -20,6 +20,9 @@ from starlette.datastructures import Headers
 from typing_extensions import TypeIs
 
 import os
+import zlib
+import numpy as np
+from typing import Tuple
 
 if sys.version_info >= (3, 12):
     from typing import TypedDict
@@ -1276,6 +1279,44 @@ class OpenAIServing:
             return self.models.base_model_paths[0].name
         return model_name
 
+    def convert_ndarray_to_str(self, data: np.ndarray) -> Tuple[str, Tuple[int, ...], str]:
+        shape = data.shape
+        dtype_str = str(data.dtype)
+        raw_bytes = data.tobytes()
+        compressed = zlib.compress(raw_bytes, level=1)
+        b64_data = base64.b64encode(compressed).decode('utf-8')
+        return b64_data, shape, dtype_str
+
+    def restore_str_to_ndarray(self, b64_data: str, shape: Tuple[int, ...], dtype_str: str) -> np.ndarray:
+        compressed = base64.b64decode(b64_data)
+        raw_bytes = zlib.decompress(compressed)
+        return np.frombuffer(raw_bytes, dtype=np.dtype(dtype_str)).reshape(shape)
+
+    def concatenate_dict_and_ndarray(self, routed_experts_dict: dict, expert_ids_arr: np.ndarray) -> np.ndarray:
+        prefill_expert_ids = routed_experts_dict.get('routed_experts_str')
+        shape = routed_experts_dict.get('routed_experts_shape', None)
+        dtype = routed_experts_dict.get('routed_experts_dtype', None)
+        if (
+            prefill_expert_ids is not None and
+            shape is not None and
+            dtype is not None and
+            len(shape) == expert_ids_arr.ndim and
+            shape[1:] == list(expert_ids_arr.shape[1:]) and
+            dtype == str(expert_ids_arr.dtype)
+        ):
+            arr = self.restore_str_to_ndarray(prefill_expert_ids, shape, dtype)
+            expert_ids_arr = np.concatenate([arr, expert_ids_arr], axis=0)
+        else:
+            logger.warning(f"{shape=} or {dtype=} mismatch {expert_ids_arr.shape=} or {str(expert_ids_arr.dtype)=}, skipping routed_experts output")
+        return expert_ids_arr
+
+    def add_ndarray_info_to_dict(self, data: np.ndarray, routed_experts_dict: dict):
+        b64_data, shape, dtype = self.convert_ndarray_to_str(data)
+        routed_experts_dict["routed_experts_shape"] = shape
+        routed_experts_dict["routed_experts_dtype"] = dtype
+        routed_experts_dict["routed_experts_str_len"] = len(b64_data)
+        routed_experts_dict["routed_experts_str"] = b64_data
+
 
 def clamp_prompt_logprobs(
     prompt_logprobs: Union[PromptLogprobs,
diff --git a/vllm/outputs.py b/vllm/outputs.py
index 33cc50c87..a35b0bb53 100644
--- a/vllm/outputs.py
+++ b/vllm/outputs.py
@@ -7,6 +7,7 @@ from dataclasses import dataclass
 from typing import Any, Generic, Optional, Union
 
 import torch
+import numpy as np
 from typing_extensions import TypeVar, deprecated
 
 from vllm.logger import init_logger
@@ -117,6 +118,7 @@ class RequestOutput:
         prompt_logprobs: Optional[PromptLogprobs],
         outputs: list[CompletionOutput],
         finished: bool,
+        routed_experts: Optional[np.ndarray] = None,
         metrics: Optional[RequestMetrics] = None,
         lora_request: Optional[LoRARequest] = None,
         encoder_prompt: Optional[str] = None,
@@ -145,6 +147,7 @@ class RequestOutput:
         self.encoder_prompt_token_ids = encoder_prompt_token_ids
         self.num_cached_tokens = num_cached_tokens
         self.kv_transfer_params = kv_transfer_params
+        self.routed_experts = routed_experts
 
     def add(self, next_output: "RequestOutput", aggregate: bool) -> None:
         """Merge subsequent RequestOutput into this one"""
diff --git a/vllm/v1/core/sched/scheduler.py b/vllm/v1/core/sched/scheduler.py
index 7038a87cd..870f132ad 100644
--- a/vllm/v1/core/sched/scheduler.py
+++ b/vllm/v1/core/sched/scheduler.py
@@ -35,6 +35,8 @@ from vllm.v1.spec_decode.metrics import SpecDecodingStats
 from vllm.v1.structured_output import StructuredOutputManager
 from vllm.v1.core.sched.request_queue import (SchedulingPolicy,
                                               create_request_queue)
+from omni.adaptors.vllm.sched.routed_experts_capturer import RoutedExpertsReader
+from omni.adaptors.vllm import envs
 
 logger = init_logger(__name__)
 
@@ -225,6 +227,16 @@ class Scheduler(SchedulerInterface):
 
                 self.cpu_npu_kv_cache_manager = CpuNpuKVCacheManager(self.kv_cache_manager, cpu_kv_cache_manager, self.max_model_len, self.enable_omni_attn)
 
+        self.routed_experts_reader = RoutedExpertsReader.create() if envs.EXPORT_MOE_EXPERTS == '1' else None
+        if self.routed_experts_reader is not None:
+            self.routed_experts_reader.attach_buffer(
+                # ,
+                # ,
+                hf_config = self.vllm_config.model_config.hf_config,
+                block_size = self.cache_config.block_size,
+                instance_id=self.vllm_config.instance_id
+            )
+
     def schedule(self) -> SchedulerOutput:
         # NOTE(woosuk) on the scheduling algorithm:
         # There's no "decoding phase" nor "prefill phase" in the scheduler.
@@ -936,6 +948,16 @@ class Scheduler(SchedulerInterface):
             # to return empty token ids for the request.
 
             stopped, num_new = add_token_and_check_stop(new_token_ids, request, self.max_model_len - (1 + self.num_spec_tokens) * num_steps + 1 if (self.vllm_config.kv_transfer_config is not None and self.vllm_config.kv_transfer_config.kv_role == "kv_consumer") else self.max_model_len)
+
+            if self.routed_experts_reader is not None:
+                    routed_experts = self.routed_experts_reader.get_routed_experts(
+                        request=request,
+                        kv_cache_manager=self.kv_cache_manager,
+                        stopped=stopped, num_new_token = num_new
+                    )
+            else:
+                routed_experts = None
+
             if stopped:
                 kv_transfer_params = self._free_request(request)
                 del new_token_ids[num_new:]  # Trim new tokens if needed.
@@ -1005,6 +1027,7 @@ class Scheduler(SchedulerInterface):
                         events=request.take_events(),
                         kv_transfer_params=kv_transfer_params,
                         num_cached_tokens=request.num_cached_tokens,
+                        routed_experts=routed_experts,
                     ))
 
             else:
diff --git a/vllm/v1/engine/__init__.py b/vllm/v1/engine/__init__.py
index 1474f9528..390107a27 100644
--- a/vllm/v1/engine/__init__.py
+++ b/vllm/v1/engine/__init__.py
@@ -6,6 +6,7 @@ from collections.abc import Sequence
 from typing import Any, Optional, Union
 
 import msgspec
+import numpy as np
 
 from vllm.lora.request import LoRARequest
 from vllm.multimodal import MultiModalKwargs
@@ -110,6 +111,7 @@ class EngineCoreOutput(
 
     # The number of tokens with prefix cache hits.
     num_cached_tokens: int = 0
+    routed_experts: Optional[np.ndarray] = None
 
     @property
     def finished(self) -> bool:
diff --git a/vllm/v1/engine/output_processor.py b/vllm/v1/engine/output_processor.py
index 293c291b4..da8aaa574 100644
--- a/vllm/v1/engine/output_processor.py
+++ b/vllm/v1/engine/output_processor.py
@@ -4,6 +4,7 @@ import asyncio
 from collections.abc import Iterable
 from dataclasses import dataclass
 from typing import Any, Optional, Union
+import numpy as np
 
 from vllm.outputs import CompletionOutput, RequestOutput
 from vllm.sampling_params import RequestOutputKind
@@ -148,6 +149,7 @@ class RequestState:
         stop_reason: Union[int, str, None],
         kv_transfer_params: Optional[dict[str, Any]] = None,
         num_cached_tokens: int = 0,
+        routed_experts: Optional[np.ndarray] = None,
     ) -> Optional[RequestOutput]:
 
         finished = finish_reason is not None
@@ -170,7 +172,7 @@ class RequestState:
                 return None
 
         return self._new_request_output(request_id, outputs, finished,
-                                        kv_transfer_params, num_cached_tokens)
+                                        kv_transfer_params, num_cached_tokens, routed_experts=routed_experts)
 
     def _new_request_output(
         self,
@@ -179,6 +181,7 @@ class RequestState:
         finished: bool,
         kv_transfer_params: Optional[dict[str, Any]] = None,
         num_cached_tokens: int = 0,
+        routed_experts: Optional[np.ndarray] = None,
     ) -> RequestOutput:
 
         if self.output_kind == RequestOutputKind.DELTA:
@@ -196,6 +199,7 @@ class RequestState:
             finished=finished,
             kv_transfer_params=kv_transfer_params,
             num_cached_tokens=num_cached_tokens,
+            routed_experts=routed_experts,
         )
 
     def _new_completion_output(
@@ -359,7 +363,7 @@ class OutputProcessor:
             # 4) Create and handle RequestOutput objects.
             if request_output := req_state.make_request_output(
                     new_token_ids, finish_reason, stop_reason,
-                    kv_transfer_params, num_cached_tokens):
+                    kv_transfer_params, num_cached_tokens, routed_experts=engine_core_output.routed_experts):
                 if req_state.queue is not None:
                     # AsyncLLM: put into queue for handling by generate().
                     req_state.queue.put(request_output)
